crypt-gost-yescrypt.c

```diff
@@ -0,0 +1,364 @@
+               "ALG_SPECIFIC_SIZE is too small for GOST-YESCRYPT.");
+
+static const uint8_t atoi64_partial[77] =
+{
+  0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,
+  64, 64, 64, 64, 64, 64, 64,
+  12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,
+  25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37,
+  64, 64, 64, 64, 64, 64,
+  38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
+  51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63
+};
+
+static const char * const itoa64 =
+  "./0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+
+static inline uint32_t atoi64(uint8_t src)
+{
+  if (src >= '.' && src <= 'z')
+    return atoi64_partial[src - '.'];
+
+  return 64;
+}
+
+static const uint8_t *decode64(uint8_t *dst, size_t *dstlen,
+                               const uint8_t *src, size_t srclen)
+{
+  size_t dstpos = 0;
+
+  while (dstpos <= *dstlen && srclen)
+    {
+      uint32_t value = 0, bits = 0;
+      while (srclen--)
+        {
+          uint32_t c = atoi64(*src);
+          if (c > 63)
+            {
+              srclen = 0;
+              break;
+            }
+          src++;
+          value |= c << bits;
+          bits += 6;
+          if (bits >= 24)
+            break;
+        }
+      if (!bits)
+        break;
+      if (bits < 12) /* must have at least one full byte */
+        goto fail;
+      while (dstpos++ < *dstlen)
+        {
+          *dst++ = (uint8_t) value;
+          value >>= 8;
+          bits -= 8;
+          if (bits < 8)   /* 2 or 4 */
+            {
+              if (value) /* must be 0 */
+                goto fail;
+              bits = 0;
+              break;
+            }
+        }
+      if (bits)
+        goto fail;
+    }
+
+  if (!srclen && dstpos <= *dstlen)
+    {
+      *dstlen = dstpos;
+      return src;
+    }
+
+fail:
+  *dstlen = 0;
+  return NULL;
+}
+
+static uint8_t *encode64_uint32_fixed(uint8_t *dst, size_t dstlen,
+                                      uint32_t src, uint32_t srcbits)
+{
+  uint32_t bits;
+
+  for (bits = 0; bits < srcbits; bits += 6)
+    {
+      if (dstlen < 2)
+        return NULL;
+      *dst++ = (uint8_t) itoa64[src & 0x3f];
+      dstlen--;
+      src >>= 6;
+    }
+
+  if (src || dstlen < 1)
+    return NULL;
+
+  *dst = 0; /* NUL terminate just in case */
+
+  return dst;
+}
+
+static uint8_t *encode64(uint8_t *dst, size_t dstlen,
+                         const uint8_t *src, size_t srclen)
+{
+  size_t i;
+
+  for (i = 0; i < srclen; )
+    {
+      uint8_t *dnext;
+      uint32_t value = 0, bits = 0;
+      do
+        {
+          value |= (uint32_t)src[i++] << bits;
+          bits += 8;
+        }
+      while (bits < 24 && i < srclen);
+      dnext = encode64_uint32_fixed (dst, dstlen, value, bits);
+      if (!dnext)
+        return NULL;
+      dstlen -= (size_t) (dnext - dst);
+      dst = dnext;
+    }
+
+  if (dstlen < 1)
+    return NULL;
+
+  *dst = 0; /* NUL terminate just in case */
+
+  return dst;
+}
```

Comment:
I'm not entirely happy about the inclusion of duplicate copies of some encoding/decoding functions from yescrypt here. I don't recall, is this how it was in @vt-alt's patch? I'm also unhappy about the coding style conversion - would be extra work to maintain both versions.

Reply:
Then it would be nice, if we those functions wouldn't be marked as `static` and added to `yescrypt.h`, so other consumers can access them easily.

Reply:
Yes, it's a matter of whether we make them part of the yescrypt API or not. Maybe it could be semi-internal API. I don't recall how @vt-alt ended up doing this, although I do recall us discussing it briefly.